# Урок 1. Расчет вероятности случайных событий

In [1]:
from math import factorial

In [3]:
# перестановка P - последовательное расположение с учетом порядка
def P(n):
    return factorial(n)

# размещение A - перестановка k элем-ов из множ-ва n
def A(n, k):
    return factorial(n) // factorial(n - k)

# размещение A - перестановка k элем-ов из множ-ва n (с повторениями)
def A_(n, k):
    return n**k

# сочетание C - кол-во способов взять k элем-ов из множ-ва n без учета порядка расположения
def C(n, k):
    return factorial(n) // (factorial(k) * (factorial(n - k)))

***
### Задание 1. В партии 10 деталей. Среди них 3 бракованные. Какова  вероятность, что среди 5, взятых на удачу, 4 хорошие детали?

Из классической формулой вероятности:\
$P(A)=\cfrac{m}{n}$ , где $m$-число **благоприятных** исходов, $n$-общее число **всех** исходов.
<br><br>

$n=C^{5}_{10}$ - общее количество исходов = способов взять 5 деталей из 10. \
$C^{4}_{7}$ - способов взять 4 детали из 7 хороших \
$C^{1}_{3}$ - способов взять 1 деталь из 3 бракованных \
Тогда $m = C^{4}_{7}\cdot C^{1}_{3}$

In [5]:
m = C(7,4) * C(3,1)
print(f'{m = }')

n = C(10,5)
print(f'{n = }')

p = m / n
print(f'P(4 хороших) = {m} / {n} = {p:.4f}')

m = 105
n = 252
P(4 хороших) = 105 / 252 = 0.4167


#### Ответ:
$P(\text{4 хороших})\approx 0.4167 \approx 41.67\%$

***
### Задание 2. Разработали спам-фильтр на основании часто встречающихся  фраз. 70% всех писем – это спам. В 10 % писем со спамом встречается фраза: «вся правда  о» и в 0.5% она встречается в хороших письмах. Какова вероятность, что пришедшее на почту письмо является спамом, если в нем есть данная фраза?

Событие $F$ - фраза «вся правда  о» в письме \
Событие $S$ - письмо-спам

Воспользуемся формулой Байеса: \
$P(S|F)=\cfrac{P(F|S)}{P(F)} \cdot P(S)$

$P(F|S) = 0.1$ - вероятность встретить фразу в спам-письме \
$P(S) = 0.7$ - вероятность того, что письмо - спам \
Вероятность встретить фразу в письме (__*по формуле полной вероятности для противоположных событий*__): \
$P(F) = P(F|S) \cdot P(S) + P(F|\bar{S}) \cdot P(\bar{S}) = 0.1 \cdot 0.7 + 0.005 \cdot (1-0.7)$

In [6]:
p = 0.1 / (0.1*0.7+0.005*0.3) * 0.7
print(f'P(S|F) = {p:.4f}')

P(S|F) = 0.9790


#### Ответ:
$P(\text{спам|фраза})\approx 0.9790 \approx 97.90\%$

***
### Задание 3. В ящике находится 10 красных, 5 черных, 5 зеленых шаров. Наудачу вынимают 6 шаров. Какова вероятность, что вынуты 3 красных, 2 черных, 1  зеленый?

Из классической формулой вероятности: \
$P(A)=\cfrac{m}{n}$, где $m$-число **благоприятных** исходов, $n$-общее число **всех** исходов.

Событие $A$ - вынули 3 красных, 2 черных, 1 зеленый

$n = C_{20}^{6}$ - общее количество исходов = способов взять 6 шаров из 20.

$m$ - складывается из: \
$C_{10}^{3}$ - способов взять 3 красных из 10 \
$C_{5}^{2}$ - способов взять 2 черных из 5 \
$C_{5}^{1}$ - способов взять 1 зеленый из 5 \
$m = C_{10}^{3} \cdot C_{5}^{2} \cdot C_{5}^{1}$ - общее количество благоприятных исходов

In [7]:
m = C(10,3) * C(5,2) * C(5,1)
print(f'{m = }')

n = C(20,6)
print(f'{n = }')

p = m / n
print(f'P(A) = {m} / {n} = {p:.4f}')

m = 6000
n = 38760
P(A) = 6000 / 38760 = 0.1548


#### Ответ:
$P(\text{A})\approx 0.1548 \approx 15.48\%$

***
### Задание 4. На 5 карточках написаны буквы:
### a) Ч, А, Й ,К, И. Какова вероятность, что получится слово ЧАЙКИ?

Событие $A$ - сложили слово ЧАЙКИ.

$n = P_{5}$ - общее количество исходов = способов перестановки букв.

$m = 1$ - только 1 способ сорбрать слово ЧАЙКИ

In [8]:
m = 1
print(f'{m = }')

n = P(5)
print(f'{n = }')

p = m / n
print(f'P(A) = {m} / {n} = {p:.4f}')

m = 1
n = 120
P(A) = 1 / 120 = 0.0083


### б) Ч, А, Й ,К, А. Какова вероятность, что получится слово ЧАЙКА?

Событие $A$ - сложили слово ЧАЙКА.

$n = P_{5}$ - общее количество исходов = способов перестановки букв.

$m = P_{1} \cdot P_{2} \cdot P_{1} \cdot P_{1}$

In [9]:
m = P(1) * P(2) * P(1) * P(1)
print(f'{m = }')

n = P(5)
print(f'{n = }')

p = m / n
print(f'P(A) = {m} / {n} = {p:.4f}')

m = 2
n = 120
P(A) = 2 / 120 = 0.0167


### в) А, Н, А, Н, А, С. Какова вероятность, что получится слово АНАНАС?

Событие $A$ - сложили слово АНАНАС.

$n = P_{6}$ - общее количество исходов = способов перестановки букв.

$m = P_{3} \cdot P_{2} \cdot P_{1}$

In [10]:
m = P(3) * P(2) * P(1)
print(f'{m = }')

n = P(6)
print(f'{n = }')

p = m / n
print(f'P(A) = {m} / {n} = {p:.4f}')

m = 12
n = 720
P(A) = 12 / 720 = 0.0167


***
### Задание 5. Какое количество семизначных номеров можно придумать, если  в качестве первой цифры не может быть 0, 1, 7?

Всего цифр 10, но на первом месте можут стоять толоко 7 из них. \
На оставшихся 6 местах может стоять любая цифра и они могут повторяться.

In [11]:
k = 7 * A_(10,6)
print(f'Кол-во номеров = {k}')

Кол-во номеров = 7000000


***
### Задание 6. Партия электрических лампочек на 25% изготовлена 1-м заводом, на 35% - 2м заводом, на 40% - третьим. Вероятность выпуска бракованых лампочек соответственно равны: $p_1=0.03$, $p_2=0.02$, $p_3=0.01$. Какова вероятность того, что взятая наудачу лампочка окажется бракованной?

По формуле полной вероятности: $\displaystyle P(A) = \sum_{i=1}^{n}{P(A|B_i)\cdot P(B_i)}$

$A$ - браковая лампочка, \
$B_1$ - изготовитель 1-й завод, $B_2$ - изготовитель 2-й завод, $B_3$ - изготовитель 3-й завод.

$P(A) = P(A|B_1) \cdot P(B_1) + P(A|B_2) \cdot P(B_2) + P(A|B_3) \cdot P(B_3) = 0.03 \cdot 0.25 + 0.02 \cdot 0.35 + 0.01 \cdot 0.4 = 0.0185$

In [13]:
print(0.03*0.25 + 0.02*0.35 + 0.01*0.4)

0.0185


***
### Задание 7. В урне a красных и b голубых шаров, одинаковых по размеру и весу. Чему равна вероятность того, что наудачу извлеченный шар из этой урны окажется голубым? Записать ответ в виде буквенного выражения.

$P(A) = \cfrac{|A|}{|\Omega|} = \cfrac{благоприятные\,исходы}{все\,исходы}$

$P(A) = \cfrac{b}{a+b}$

***
### Задание 8. Партия деталей изготовлена тремя рабочими, причем первый изготовил 35% всех деталей, второй - 40% всех деталей, третий - всю остальную продукцию. Брак в их продукции составляет: у первого - 2%, у второго - 3%, у третьего - 4%. Случайно выбранная для контроля деталь оказалась бракованной. Найти вероятность того, что она изготовлена третьим рабочим.

По теореме Байеса: $P(A|B) = \cfrac{P(B|A)}{P(B)} \cdot P(A)$

По формуле полной вероятности: $\displaystyle P(A) = \sum_{i=1}^{n}{P(A|B_i) \cdot P(B_i)}$
<br><br>

$A$ - браковая деталь, \
$B_1$ - сделал 1-й рабочий, $B_2$ - сделал 2-й рабочий, $B_3$ - сделал 2-й рабочий.

Найти $P(A|B_3) = \cfrac{P(B_3|A)}{P(B_3)} \cdot P(A)$

$P(B_3) = 1 - (0.35 + 0.40) = 0.25$ \
$P(B_3|A) = 0.04$ \
$P(A) = P(A|B_1) \cdot P(B_1) + P(A|B_2) \cdot P(B_2) + P(A|B_3) \cdot P(B_3) = 0.02 \cdot 0.35 + 0.03 \cdot 0.4 + 0.04 \cdot 0.25 = 0.029$

$P(A|B_3) = \cfrac{0.04}{0.25} \cdot 0.029 = 0.00464$

In [14]:
print(0.02*0.35 + 0.03*0.4 + 0.04*0.25)
print(0.04/0.25*0.029)

0.028999999999999998
0.00464


***
### Задание 9. В ящике 15 шаров, из которых 5 голубы и 10 красных. Из ящика последовательно вынимают 2 шара; первый шар в ящик не возвращают. Найти вероятность, что первый вытащенный шар - красный, а второй – голубой.

$P(KG) = P(K) \cdot P(G) = \cfrac{10}{15} \cdot \cfrac{5}{14} = 0.2381$

In [15]:
print(10/15 * 5/14)

0.23809523809523808
